File: go/internal/clock/vector_clock_test.go
package clock

import (
	"testing"
)

func TestIncrement(t *testing.T) {
	clock := NewVectorClock()
	clock = Increment(clock, "peer1")
	if clock["peer1"] != 1 {
		t.Errorf("Expected 1, got %d", clock["peer1"])
	}
	clock = Increment(clock, "peer1")
	if clock["peer1"] != 2 {
		t.Errorf("Expected 2, got %d", clock["peer1"])
	}
}

func TestIncrementNil(t *testing.T) {
	var clock VectorClock
	clock = Increment(clock, "peer1")
	if clock["peer1"] != 1 {
		t.Errorf("Expected 1, got %d", clock["peer1"])
	}
}

func TestMerge(t *testing.T) {
	clock1 := VectorClock{"a": 1, "b": 2}
	clock2 := VectorClock{"a": 3, "c": 4}
	merged := Merge(clock1, clock2)
	if merged["a"] != 3 || merged["b"] != 2 || merged["c"] != 4 {
		t.Errorf("Merge failed: %v", merged)
	}
}

func TestCompare(t *testing.T) {
	clock1 := VectorClock{"a": 1, "b": 2}
	clock2 := VectorClock{"a": 1, "b": 2}
	if Compare(clock1, clock2) != Equal {
		t.Error("Expected Equal")
	}

	clock3 := VectorClock{"a": 2, "b": 2}
	if Compare(clock1, clock3) != Before {
		t.Error("Expected Before")
	}

	clock4 := VectorClock{"a": 0, "b": 2}
	if Compare(clock1, clock4) != After {
		t.Error("Expected After")
	}

	clock5 := VectorClock{"a": 2, "b": 1}
	if Compare(clock1, clock5) != Concurrent {
		t.Error("Expected Concurrent")
	}
}

func TestHappensBefore(t *testing.T) {
	clock1 := VectorClock{"a": 1, "b": 2}
	clock2 := VectorClock{"a": 1, "b": 2}
	if !HappensBefore(clock1, clock2) {
		t.Error("Equal should happen before")
	}

	clock3 := VectorClock{"a": 2, "b": 2}
	if !HappensBefore(clock1, clock3) {
		t.Error("Before should happen before")
	}

	clock4 := VectorClock{"a": 0, "b": 2}
	if HappensBefore(clock1, clock4) {
		t.Error("After should not happen before")
	}
}

func TestClone(t *testing.T) {
	clock := VectorClock{"a": 1, "b": 2}
	cloned := Clone(clock)
	if cloned["a"] != 1 || cloned["b"] != 2 {
		t.Errorf("Clone failed: %v", cloned)
	}
	cloned["a"] = 3
	if clock["a"] != 1 {
		t.Error("Clone should be independent")
	}
}

func TestCloneNil(t *testing.T) {
	var clock VectorClock
	cloned := Clone(clock)
	if cloned != nil {
		t.Error("Clone of nil should be nil")
	}
}

File: ts/src/components/clock/vector_clock.test.ts
import { increment, merge, compare, ComparisonResult, happensBefore, newVectorClock, clone } from './vector_clock';

describe('VectorClock', () => {
  describe('increment', () => {
    it('should increment existing peer counter', () => {
      const clock = { a: 1, b: 2 };
      const result = increment(clock, 'a');
      expect(result.a).toBe(2);
      expect(result.b).toBe(2);
    });

    it('should add new peer counter', () => {
      const clock = { a: 1 };
      const result = increment(clock, 'b');
      expect(result.a).toBe(1);
      expect(result.b).toBe(1);
    });

    it('should handle null clock', () => {
      const result = increment(null as any, 'a');
      expect(result.a).toBe(1);
    });
  });

  describe('merge', () => {
    it('should merge two clocks taking max values', () => {
      const clock1 = { a: 1, b: 2 };
      const clock2 = { a: 3, c: 4 };
      const result = merge(clock1, clock2);
      expect(result).toEqual({ a: 3, b: 2, c: 4 });
    });
  });

  describe('compare', () => {
    it('should return Equal for identical clocks', () => {
      const clock1 = { a: 1, b: 2 };
      const clock2 = { a: 1, b: 2 };
      expect(compare(clock1, clock2)).toBe(ComparisonResult.Equal);
    });

    it('should return After when first clock is ahead', () => {
      const clock1 = { a: 2, b: 2 };
      const clock2 = { a: 1, b: 2 };
      expect(compare(clock1, clock2)).toBe(ComparisonResult.After);
    });

    it('should return Before when first clock is behind', () => {
      const clock1 = { a: 1, b: 2 };
      const clock2 = { a: 2, b: 2 };
      expect(compare(clock1, clock2)).toBe(ComparisonResult.Before);
    });

    it('should return Concurrent for concurrent clocks', () => {
      const clock1 = { a: 2, b: 1 };
      const clock2 = { a: 1, b: 2 };
      expect(compare(clock1, clock2)).toBe(ComparisonResult.Concurrent);
    });
  });

  describe('happensBefore', () => {
    it('should return true for equal clocks', () => {
      const clock1 = { a: 1, b: 2 };
      const clock2 = { a: 1, b: 2 };
      expect(happensBefore(clock1, clock2)).toBe(true);
    });

    it('should return true for before clocks', () => {
      const clock1 = { a: 1, b: 2 };
      const clock2 = { a: 2, b: 2 };
      expect(happensBefore(clock1, clock2)).toBe(true);
    });

    it('should return false for after clocks', () => {
      const clock1 = { a: 2, b: 2 };
      const clock2 = { a: 1, b: 2 };
      expect(happensBefore(clock1, clock2)).toBe(false);
    });
  });

  describe('newVectorClock', () => {
    it('should return empty object', () => {
      const clock = newVectorClock();
      expect(clock).toEqual({});
    });
  });

  describe('clone', () => {
    it('should return shallow copy', () => {
      const original = { a: 1, b: 2 };
      const cloned = clone(original);
      expect(cloned).toEqual(original);
      expect(cloned).not.toBe(original); // Different reference
    });

    it('should handle null clock', () => {
      const cloned = clone(null as any);
      expect(cloned).toEqual({});
    });
  });
});